File: leetcode/1-easy/9-other/3-reversebits/reverse_bits.c
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/99/others/648/
// https://leetcode.com/problems/reverse-bits/description/

// Reverse Bits
//
// Reverse bits of a given 32 bits unsigned integer.

// this is not flip the bits, but rather, reverse the bits (like reverse string)

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>


// 600 / 600 test cases passed.
// Runtime: 12 ms
// only beats 20.73% of C submission
uint32_t reverseBits(uint32_t n) {

    // this is what I came up with. Sort of based on the short-circuit
    // hamming weight calculation.
    
    uint32_t v = 0;
    uint32_t mask = (1 << 31);  // set MSB to 1
    while (n) {  // while n have some bits on
        if (n & 1) {  // if LSB bit is on
            v |= mask;
        }
        n >>= 1;
        mask >>= 1;
    }
    return v;
}


// 600 / 600 test cases passed.
// Runtime: 4 ms
// beats 86.59% of C submission
//
uint32_t reverseBits2(uint32_t n) {

    // From forum post. It's interesting that this is 3 times faster
    // than above. not obvious from the code. Perhaps the compiler can
    // optimize this better?

    uint32_t v = 0;
    for (int i = 0; i < 32; i++, n >>= 1) {
        v <<= 1;
        v |= (n & 1);  // set v's LSB to on if n's LSB is on
    }
    return v;
}


int main(int argc, char *argv[], char *env[]) {

    if (argc < 2) {
        printf("Usage: %s <n>\n", argv[0]);
        return 1;
    }

    uint32_t n = atoi(argv[1]);
    // uint32_t v = reverseBits(n);
    uint32_t v = reverseBits2(n);

    printf("%u: %u\n", n, v);

    return 0;
}

File: leetcode/1-easy/9-other/3-reversebits/reverse_bits.py
# https://leetcode.com/explore/interview/card/top-interview-questions-easy/99/others/648/
# https://leetcode.com/problems/reverse-bits/description/

# Reverse Bits
#
# Reverse bits of a given 32 bits unsigned integer.

# this is not flip the bits, but rather, reverse the bits (like reverse string)

import unittest


class Solution:

    # @param n, an integer
    # @return an integer
    def reverseBits(self, n):
        return self.my_reverse2(n)

    # Note: for each bit in n, set bit in m and shift left
    #
    # process n from right (LSB), set m from right and shift result
    # left.

    # reverse via bit operation
    def my_reverse2(self, n):
        m = 0
        for i in range(32):
            m <<= 1  # shift current result to left
            m |= n & 1  # set LSB on m
            n >>= 1  # shift input to right
        return m

    # 600 / 600 test cases passed.
    # 35ms
    # beats 99.32% of python submission
    def my_reverse(self, n):
        # bin(n)[2:]       # convert to bin, chop off '0b' prefix
        # .rjust(32, '0')  # fill left with 0 until width is 32
        # int(x[::-1], 2)  # reverse, then convert int from base2
        return int(bin(n)[2:].rjust(32, '0')[::-1], 2)


class TestReverse(unittest.TestCase):

    def setUp(self):
        self.sol = Solution()

    def test1(self):
        print(self.sol.reverseBits(123))


if __name__ == '__main__':
    unittest.main()
